Block patterns: switch to user locale and allow all core/Gutenberg pattern operations

ramonjd opened the pull request:
#### Changes proposed in this Pull Request

This PR:

- Change the locale to the user locale to make it consistent with core patterns
- Pass site/block editor string to Block_Patterns_From_API. Let's let the Block_Patterns_From_API class house the logic as to which patterns source to load.
- Remove all methods that register/unregister Gutenberg or core patterns. Instead we'll let Gutenberg do it's thing (See: https://github.com/WordPress/gutenberg/blob/trunk/lib/block-patterns.php )
- Updated tests.


#### Testing instructions

- [ ] Turn on some relaxing background music.
- [ ] Apply this PR's changes to your sandbox: `install-plugin.sh editing-toolkit update/block-patterns-user-locale-and-allow-gutenberg-patterns` and sandbox a test site using a block theme e.g., TT1 Blocks
- [ ] Now fire up the block editor. 
- [ ] Check that the query blocks are appearing (Gutenberg registers these in [block-patterns.php](https://github.com/WordPress/gutenberg/blob/trunk/lib/block-patterns.php))
<img width="150" alt="Screen Shot 2021-07-26 at 11 57 18 am" src="https://user-images.githubusercontent.com/6458278/126923123-d42178bc-fb80-40f7-9f59-c09fae6007d5.png">

- [ ] Check that other Gutenberg blocks are appearing. There's only one for now: `Social links with a shared background color`

<img width="150" alt="Screen Shot 2021-07-26 at 11 50 29 am" src="https://user-images.githubusercontent.com/6458278/126923142-dfc8c0e0-090a-4292-9199-aac37da4e47b.png">

- [ ] Check that Wordpress core patterns are being loaded from the [patterns directory](https://wordpress.org/patterns/categories/columns/?orderby=favorite_count) and **not** from [WordPress core itself](https://github.com/WordPress/WordPress/blob/master/wp-includes/block-patterns.php). (Gutenberg unregisters core patterns in [block-patterns.php](https://github.com/WordPress/gutenberg/blob/trunk/lib/block-patterns.php))

For example, here's one in the columns category:

<img width="150" alt="Screen Shot 2021-07-26 at 11 50 21 am" src="https://user-images.githubusercontent.com/6458278/126923143-f8357a97-b345-462b-9363-a6babef53af5.png">

Here's the source preview from the patterns directory

<img width="150" alt="Screen Shot 2021-07-26 at 11 51 04 am" src="https://user-images.githubusercontent.com/6458278/126923140-8774b04e-5646-4767-888b-20be5d113144.png">

- [ ] Check that WPCOM patterns are being loaded from [dotcom patterns](https://dotcompatterns.wordpress.com/2021/05/07/gallery/). Here's one in French!
<img width="150" alt="Screen Shot 2021-07-26 at 11 54 34 am" src="https://user-images.githubusercontent.com/6458278/126923132-3018aa2c-cde2-42b4-8a03-d81b0d93f363.png">

- [ ] Switch your **user language**. To test translations from the patterns directory, like many other things, it's best to go with Japanese. Make sure the patterns are coming through translated.

<img width="150" alt="Screen Shot 2021-07-26 at 11 56 50 am" src="https://user-images.githubusercontent.com/6458278/126923131-fde3c8b1-9dbb-4584-b39c-1be6ae9b8e83.png">

<img width="150" alt="Screen Shot 2021-07-26 at 11 52 12 am" src="https://user-images.githubusercontent.com/6458278/126923139-c66c8b45-5fdf-4d19-a99b-c30fbacce33c.png">

- [ ] Head over to the site editor. Ensure that things work in the same way.
- [ ] Run unit tests

```bash
cd apps/editing-toolkit
yarn run test:php --testsuite block-patterns
```

- [ ] Congratulate yourself on a job well done!



Related to https://github.com/Automattic/wp-calypso/pull/53243

ramonjd committed:
Change the locale to the user locale to make it consistent with core patterns
Pass site/block editor string to Block_Patterns_From_API. Let's let the Block_Patterns_From_API class house the logic as to which patterns source to load.
Remove all methods that register/unregister Gutenberg or core patterns. Instead we'll let Gutenberg do it's thing
Updated tests.

ramonjd commented:
@andrewserong If you get a chance this week, would you mind running your learned eye over this one? Thank you 🙇

andrewserong commented:
Thanks for working on this one @ramonjd, yes, will be good to get this one in this week. I'm still catching up on a few things after AFK, but if I don't get to this today, I'll be sure to give it a 👀 tomorrow!

andrewserong commented on apps/editing-toolkit/editing-toolkit-plugin/block-patterns/test/class-block-patterns-from-api-test.php:
```diff
@@ -107,12 +107,42 @@ public function test_patterns_request_succeeds_with_empty_cache() {
+			return true;
+		};
+
+		add_filter( 'a8c_enable_fse_block_patterns_api', $should_enable_fse_block_patterns_api );
```
Tiny nit, and totally optional, but could we reduce a few lines by using [__return_true](https://developer.wordpress.org/reference/functions/__return_true/) as the callback?

ramonjd replied:
Works for me! No harm in removing unnecessary lines :)

andrewserong commented on apps/editing-toolkit/editing-toolkit-plugin/block-patterns/class-block-patterns-from-api.php:
```diff
@@ -53,16 +38,30 @@ class Block_Patterns_From_API {
-		$patterns_sources       = empty( $patterns_sources ) ? array( 'block_patterns' ) : $patterns_sources;
-		$this->patterns_sources = empty( array_diff( $patterns_sources, $this->valid_patterns_sources ) ) ? $patterns_sources : array( 'block_patterns' );
-		$this->utils            = empty( $utils ) ? new \A8C\FSE\Block_Patterns_Utils() : $utils;
+	public function __construct( string $editor_type = 'block_editor', Block_Patterns_Utils $utils = null ) {
```
Just wondering about potential mid-deploy fatals with this change to the constructor. E.g. if during deployment this file is changed before `full-site-editing-plugin.php`, then this function _might_ receive an array instead of a string in its first param? Would it be worth dropping the `string` type hint and adding an `is_array` check in the constructor to handle the possibility of an array, and then in a follow-up remove the `is_array` check and add back in the type hint (along with deploying the change to `full-site-editing-plugin.php`)?

ramonjd replied:
Good spotting. Best to be sure here. I hadn't considered that. 

What do you think about a check like:

```php
		if ( ! is_string( $editor_type ) ) {
			$editor_type = 'block_editor';
		}

```

?

andrewserong reviewed (approved):
This is testing nicely for me @ramonjd and there's a good feeling of consistency now that we're matching core's behaviour of using the user locale. Thanks for cleaning this up!

![image](https://user-images.githubusercontent.com/14988353/126949729-60601e89-5524-48bc-ae6f-e3f53e30d49a.png)

✅ Turn on some relaxing background music. [Done](https://open.spotify.com/track/5u5aVJKjSMJr4zesMPz7bL?si=e6d33efbc2f94bce)!
✅ Check that the query blocks are appearing (Gutenberg registers these in block-patterns.php)
✅ Check that other Gutenberg blocks are appearing. There's only one for now: Social links with a shared background color
✅ Check that Wordpress core patterns are being loaded from the patterns directory and not from WordPress core itself
✅ Check that WPCOM patterns are being loaded from dotcom patterns
✅ Switch your user language. Make sure the patterns are coming through translated.
✅ Head over to the site editor. Ensure that things work in the same way.

The code change looks good to go to me, so I've approved. But, I've left a comment about a proposed change to the constructor to potentially help avoid mid-deploy fatals. Might be overkill, but I wasn't _too_ sure the risk of changing the constructor signature with how this gets deployed. 🤔🤞

ramonjd committed:
- Defensive check in the Block_Patterns_From_API constructor for deployment, just in case an array (the former first argument) slips through. We can revert this back to a strongly typed variable after deployment
- in the unit test use the WordPress `__return_true` callback shorthand in the hook.

ramonjd commented:
Appreciate the quick review @andrewserong ! Suggestions implemented 👍

andrewserong commented:
Thanks for implementing those changes @ramonjd, this looks good to me now! Just did a quick smoke test, too, to ensure it's still working correctly 👍
